server/pbench/bin/pbench-inotify

```diff
@@ -0,0 +1,227 @@
+#!/usr/bin/env python3
+"""
+This script is a generic inotify script running as a systemd
+service. Using this script, the various server scripts can be invoked
+directly and immediately, instead of having to wait to be invoked by
+cron.
+
+Not every server script will be converted: some will continue to
+be run as cron jobs.
+
+The script uses the pyinotify library to create a watch for a file and
+triggers the associationed command if any changes happen to the
+file. IOW, we assume that a new entry is added to the file when a user
+submits a new result (i.e.  pbench-move-results on the agent side
+needs to be modified).
+"""
+
+import pyinotify
+import subprocess
+import os
+import sys
+import signal
+import fcntl
+import threading
+
+DEBUG = "False"
+
+class Association(object):
+
+    """
+    The Association class associates a server script to the administrative
+    files that are used to figure out whether the script needs to be
+    called.
+
+    The list_file is modified externally(e.g. remotely by
+    pbench-move-results). At initialization, the inotify handler
+    establishes watches on *all* the list_files, so when a change is seen,
+    it use the file name of the changed file as a key in a dict of
+    associations to select the correct one and calls the
+    compare_and_dispatch() routine for this association.
+
+    compare_and_dispatch() calculates the difference between the list_file
+    and the(in-memory) done_links list, to figure out the list of new
+    results. It then calls the script for this association, passing it
+    that list of new results.
+
+    Periodically during cleanup, the list_file is appended to old_list_file and the done_file along with the done_links list is appended to the old_done_file.
+
+    And also when this script is terminated, done_links list which is in memory
+    is appended to done_file to save the state of the script.
```

Comment:
It would help to have an explanation of the various states of `done_links`:  something along the lines of
 
> done_links is an in-memory list of symlinks (to tarballs). When a tarball is processed, the link is added to done_links. If the script terminates for whatever reason, the list is saved in a file (done_file). When the script is (re)started, done_links is initialized from the done_file.

But make sure that the description is correct: I'm not sure that what I wrote above *is* correct.